### PR TITLE
Add driving heuristics and introduce operations

### PR DESCRIPTION
**Pull request by dynbit**

### CODE REVIEW

File: VelocitySDK/Core/VLTRecordingConfig.m [hunk 1]
```diff
@@ -18,17 +19,30 @@ @interface VLTRecordingConfig ()
...
+
+    if (sampleSize > 0 && captureInterval > 0) {
+        VLTRecordingConfig *config = [[VLTRecordingConfig alloc] initSampleSize:sampleSize
+                                                                       interval:captureInterval
+                                                               detectioMotionOn:detectMotionOn];
+        return config;
     }
-    return self;
+
+    if (error != NULL) {
```
Comment (Antanas) [inline VelocitySDK/Core/VLTRecordingConfig.m:35]: ```if (error != nil) ``` or just ``` if (error) ```

File: VelocitySDK/Operations/DetectOperations/VLTDrivingDetectOperation.m [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    if (gpsSensorIndex == NSNotFound) {
+        return;
+    }
+
+    VLTPBSensor *gpsSensor = sensors[gpsSensorIndex];
+
+
+    Float32 sum = 0;
+    for (VLTPBSample *sample in gpsSensor.samples) {
```
Comment (Antanas) [inline VelocitySDK/Operations/DetectOperations/VLTDrivingDetectOperation.m:43]: Just to check - locations have speed "-1" when they are invalid. Are you ignoring them ?

Are gps samples removed after detecting driving/parking?

File: VelocitySDK/Operations/DetectOperations/VLTDrivingDetectOperation.m [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    self.averageSpeed = sum/gpsSensor.samples.count;
+    self.isDriving = self.averageSpeed >= self.speedThreshold;
+}
+
+- (Float32)speedFromSample:(VLTPBSample *)sample
+{
+    if (sample.values.count > VLTPBSampleSpeedIndex) {
+        return [sample valuesAtIndex:VLTPBSampleSpeedIndex];
+    } else {
+        return 0.0;
```
Comment (Antanas) [inline VelocitySDK/Operations/DetectOperations/VLTDrivingDetectOperation.m]: Just return 0.0 without else

File: VelocitySDK/Operations/VLTMotionTrackingOperation.m [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+- (BOOL)isAsynchronous
+{
+    return YES;
+}
+
+- (void)processMotionRequest
+{
+//    vlt_weakify(self);
+//    [[VLTApiClient shared] uploadForTracking:self.motionRequest
```
Comment (Antanas) [inline VelocitySDK/Operations/VLTMotionTrackingOperation.m]: Is this needed?

File: VelocitySDK/Sensors/Samples/VLTSampleBuilder.m [hunk 5]
```diff
@@ -13,6 +13,12 @@
 #import "VLTMacros.h"
 #import <CoreLocation/CoreLocation.h>
 
+const NSTimeInterval VLTPBSampleLatitudeIndex = 0;
+const NSTimeInterval VLTPBSampleLongitudeIndex = 1;
+const NSTimeInterval VLTPBSampleAltitudeIndex = 2;
+const NSTimeInterval VLTPBSampleHorizontalAccuracyIndex = 3;
+const NSTimeInterval VLTPBSampleSpeedIndex = 4;
```
Comment (Antanas) [inline VelocitySDK/Sensors/Samples/VLTSampleBuilder.m]: Maybe ```NSInteger``` fit here better then ```NSTimeInterval``` ?

File: VelocitySDK/Support/VLTMacros.h [hunk 6]
```diff
@@ -20,7 +20,7 @@ __strong typeof(var) var = weak_##var; \
 _Pragma("clang diagnostic pop")
 
 
-#define DEBUG_LOG 0
+#define DEBUG_LOG 1
```
Comment (Antanas) [inline VelocitySDK/Support/VLTMacros.h]: This shouldn't be commited

File: VelocitySDK/VLTClient.m [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+@property (atomic, assign, getter=isInProgress) BOOL inProgress;
+@property (atomic, assign) UInt32 sequenceIndex;
+
+@property (nonatomic, strong) NSOperationQueue *processingQueue;
+
+@end
+
+@implementation VLTClient
+
+@synthesize active = _active;
```
Comment (Antanas) [inline VelocitySDK/VLTClient.m:39]: Is this needed? I thought all properties are synthesized automatically

File: VelocitySDK/VLTErrors.h [hunk 8]
```diff
@@ -14,7 +14,8 @@ enum : NSInteger {
     VLTApiFatalError                        = 1001, // Fatal api error. Stops further motions sensing until app is restarted
     VLTApiTokenNotRecognized                = 1002, // Api token is not recognized. Stops further motions sensing until app is restarted
     VLTApiTokenNoAccess                     = 1003, // Api token doesn't have access or was revoked. Stops further motions sensing until app is restarted
-    VLTPayloadParseFailed                   = 1004, // Payload parsing failed. No actions should be taken
+    VLTParseError                           = 1004, // Payload parsing failed. No actions should be taken
+    VLTConfigurationError                   = 1005,
```
Comment (Antanas) [inline VelocitySDK/VLTErrors.h]: Small explanation would be great as all others has one

**Review by Antanas — APPROVED** [review]

File: VelocitySDK/VLTClient.m (hunk 7 above)
Comment (dynbit) [reply VelocitySDK/VLTClient.m:39]: afaik, if you override both - setter & getter compiler throw an error that e.g. `_active` is undeclared identifier.

File: VelocitySDK/Core/VLTRecordingConfig.m (hunk 1 above)
Comment (dynbit) [reply VelocitySDK/Core/VLTRecordingConfig.m:35]: Yours will work, although from types perspective, it should be a comparison with `NULL`.

File: VelocitySDK/Operations/DetectOperations/VLTDrivingDetectOperation.m (hunk 2 above)
Comment (dynbit) [reply VelocitySDK/Operations/DetectOperations/VLTDrivingDetectOperation.m:43]: Good catch. And also looked to the docs 
> Negative if speed is invalid.

But it doesn't say that location it self is invalid. Just the speed.

Comment (dynbit) [reply VelocitySDK/Operations/DetectOperations/VLTDrivingDetectOperation.m:43]: And answer to last questions - nop, we keep last x sample in the buffer.

File: VelocitySDK/Sensors/Samples/VLTSampleBuilder.m (hunk 5 above)
Comment (dynbit) [reply VelocitySDK/Sensors/Samples/VLTSampleBuilder.m]: 👍

**Pull request merged**